compiler/rustc_middle/src/middle/stability.rs

```diff
@@ -597,4 +598,90 @@ impl<'tcx> TyCtxt<'tcx> {
     pub fn lookup_deprecation(self, id: DefId) -> Option<Deprecation> {
         self.lookup_deprecation_entry(id).map(|depr| depr.attr)
     }
+
+    /// Returns true if `def_id` has an attribute that allows usage of the const unstable feature `feature_gate`.
+    pub fn rustc_allow_const_fn_unstable(self, def_id: LocalDefId, feature_gate: Symbol) -> bool {
+        let attrs = self.hir().attrs(self.local_def_id_to_hir_id(def_id));
+        attr::rustc_allow_const_fn_unstable(self.sess, attrs).any(|name| name == feature_gate)
+    }
+
+    pub fn enforce_trait_const_stability(
+        self,
+        trait_def_id: DefId,
+        span: Span,
+        parent_def: Option<LocalDefId>,
+    ) {
+        let Some(ConstStability {
+            level: attr::StabilityLevel::Unstable { implied_by: implied_feature, .. },
+            feature,
+            ..
+        }) = self.lookup_const_stability(trait_def_id)
+        else {
+            return;
+        };
+
+        let unstable_feature_allowed = span.allows_unstable(feature)
+            || implied_feature.is_some_and(|f| span.allows_unstable(f));
+
+        let feature_enabled = trait_def_id.is_local()
+            || self.features().enabled(feature)
+            || implied_feature.is_some_and(|f| self.features().enabled(f));
+
+        if !unstable_feature_allowed && !feature_enabled {
+            let mut diag = self.dcx().create_err(crate::error::UnstableConstTrait {
+                span,
+                def_path: self.def_path_str(trait_def_id),
+            });
+            self.disabled_nightly_features(&mut diag, None, [(String::new(), feature)]);
+            diag.emit();
+        } else if let Some(parent) = parent_def {
```

Comment:
I don't understand this -- what is the "parent" here?

The equivalent code in check.rs is
```rust
                        if !feature_enabled || !callee_safe_to_expose_on_stable {
                            self.check_op(ops::FnCallUnstable {
                                def_id: callee,
                                feature,
                                feature_enabled,
                                safe_to_expose_on_stable: callee_safe_to_expose_on_stable,
                            });
                        }
```
Here `callee_safe_to_expose_on_stable` is always `false` so we should always behave as-if `check_op` was called.

Reply:
Either a `const fn` that contains the `~const Tr` bound or the callee of a const method.



> we should always behave as-if `check_op` was called.

Yes, this is why this code is derived from `check_op_spanned`.

Reply:
> Either a const fn that contains the ~const Tr bound or the callee of a const method.

Sorry I don't follow.

The existing const checks have nothing like this, as far as I can see. I have no idea what it means here that we are skipping the check if there is no parent.

Reply:
> Yes, this is why this code is derived from check_op_spanned.

Can you make the structure parallel, i.e. also have a function like that here with essentially the same inputs (without the `NonConstOp` trait of course). Or ideally share that code rather than duplicate it... maybe we need a `compiler/rustc_middle/src/middle/const_stability.rs` file for those functions. (`stability.rs` is already huge.)